### DISCUSSION TITLE
Add a new XML based language which uses .xml extension

### DISCUSSION THREAD

**Discussion by mdaloia** [github-linguist/linguist#5922] [Classification] [answered]
Hi!

I would like to contribute a new language (`Mule`) in order to be shown in GitHub statistics of repositories.

Mule apps are written using XML language but unfortunately they don't have a specific file extension and uses `.xml`. But they must have a root node called `<mule>` with a specific namespace `xmlns="http://www.mulesoft.org/schema/mule/core"`.

After reading the docs I want to confirm if the following steps are the correct ones:

1. Add a new entry in `languages.yml` (mostly based on the `XML` entry).
2. Add a new entry in `heuristics.yml` for extension `.xml` with a matching pattern.
3. Add some samples under `samples/Mule/` directory.
4. Run `script/update-ids`
5. Run `bundle exec rake samples` to compile samples.
6. Run tests with `bundle exec rake`

As the grammar and Ace/CodeMirror modes will be the same as for `XML` language then there is no need to perform the steps to add a new one.

**Questions**:

1. Am I missing something?
2. Should I need to request a performance test of the Bayesian classifier?
3. Is there any way to know which of strategy was the "winning" one? (I mean, the one whose result was just one language)

Thanks in advance
Martin

**Comment by lildude** [comment, marked as answer]
> * Am I missing something?

Nope, that's pretty much what's documented in the CONTRIBUTING.md file for adding a language.

> * Should I need to request a performance test of the Bayesian classifier?

No need. You also shouldn't get this far as you're implementing a heuristic which should catch things before they fall through to the classifier.

> * Is there any way to know which of strategy was the "winning" one? (I mean, the one whose result was just one language)

Sort of. As you're using an existing extension your heuristic should catch things. You can confirm this by adding a test with all the other heuristic tests. You can also see that things don't fall through to the classifier by using `LINGUIST_DEBUG=1 bundle exec bin/github-linguist <path to repo>`. You'll know things haven't hit the classifier as you _won't_ see any of the token break down etc that this option reveals.
